File: app/src/main/java/net/nonylene/photolinkviewer/TwitterDisplay.java
package net.nonylene.photolinkviewer;

import android.app.Activity;
import android.app.FragmentTransaction;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteException;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import java.net.MalformedURLException;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import twitter4j.AsyncTwitter;
import twitter4j.MediaEntity;
import twitter4j.Status;
import twitter4j.TwitterAdapter;
import twitter4j.TwitterException;
import twitter4j.TwitterListener;
import twitter4j.TwitterMethod;


public class TwitterDisplay extends Activity {
    private AsyncTwitter twitter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_twitter_display);
        // get intent and purse url
        SharedPreferences sharedPreferences = getSharedPreferences("preference", Context.MODE_PRIVATE);
        if (Intent.ACTION_VIEW.equals(getIntent().getAction())) {
            Bundle bundle = new Bundle();
            Uri uri = getIntent().getData();
            String url = uri.toString();
            bundle.putString("url", url);
            // option fragment
            FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
            OptionFragment optionFragment = new OptionFragment();
            optionFragment.setArguments(bundle);
            fragmentTransaction.add(R.id.root_layout, optionFragment).commit();
            if (url.contains("twitter.com")) {
                Log.v("twitter", url);
                Pattern pattern = Pattern.compile("^https?://twitter\\.com/\\w+/status[es]*/(\\d+)");
                Matcher matcher = pattern.matcher(url);
                if (matcher.find()) {
                    Log.v("match", "success");
                }
                String id = matcher.group(1);
                Long id_long = Long.parseLong(id);
                if (sharedPreferences.getBoolean("authorized", false)) {
                    // oAuthed
                    try {
                        // get twitter async
                        twitter = MyAsyncTwitter.getAsyncTwitter(getApplicationContext());
                        twitter.addListener(twitterListener);
                        twitter.showStatus(id_long);
                        bundle.putLong("id_long", id_long);
                        TwitterOptionFragment twitterOptionFragment = new TwitterOptionFragment();
                        twitterOptionFragment.setArguments(bundle);
                        getFragmentManager().beginTransaction().add(R.id.root_layout, twitterOptionFragment).commit();
                    } catch (SQLiteException e) {
                        Log.e("SQL", e.toString());
                    }
                } else {
                    Toast.makeText(getApplicationContext(), getString(R.string.twitter_display_oauth), Toast.LENGTH_LONG).show();
                    Intent intent = new Intent(this, TOAuth.class);
                    startActivity(intent);
                }
            }
        } else {
            Toast.makeText(this, "Intent Error!", Toast.LENGTH_LONG).show();
        }
    }

    private TwitterListener twitterListener = new TwitterAdapter() {

        @Override
        public void onException(TwitterException e, TwitterMethod twitterMethod) {
            Log.e("twitterException", e.toString());
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(getApplicationContext(), getString(R.string.twitter_error_toast), Toast.LENGTH_LONG).show();
                }
            });
        }

        @Override
        public void gotShowStatus(final Status status) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    // set media entity
                    MediaEntity[] mediaEntities = status.getExtendedMediaEntities();

                    // if number of media entity is one, show fragment directly
                    if (mediaEntities.length == 1) {
                        Bundle bundle = new Bundle();
                        bundle.putString("url", mediaEntities[0].getMediaURL());
                        FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
                        ShowFragment showFragment = new ShowFragment();
                        showFragment.setArguments(bundle);
                        fragmentTransaction.replace(R.id.show_frag_replace, showFragment).commit();
                    } else {
                        ScrollView scrollView = (ScrollView) findViewById(R.id.twitterScrollView);
                        // transparent to f5
                        scrollView.setBackgroundColor(Color.parseColor("#F5F5F5"));
                        // put status on text
                        TextView textView = (TextView) findViewById(R.id.twTxt);
                        TextView snView = (TextView) findViewById(R.id.twSN);
                        TextView dayView = (TextView) findViewById(R.id.twDay);
                        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
                        URL iconUrl;

                        try {
                            //retweet check
                            if (status.isRetweet()) {
                                Status retweetedStatus = status.getRetweetedStatus();
                                textView.setText(retweetedStatus.getText());
                                snView.setText(retweetedStatus.getUser().getScreenName());
                                String statusDate = dateFormat.format(retweetedStatus.getCreatedAt());
                                dayView.setText(statusDate);
                                iconUrl = new URL(retweetedStatus.getUser().getBiggerProfileImageURL());
                            } else {
                                textView.setText(status.getText());
                                snView.setText(status.getUser().getScreenName());
                                String statusDate = dateFormat.format(status.getCreatedAt());
                                dayView.setText(statusDate);
                                iconUrl = new URL(status.getUser().getBiggerProfileImageURL());
                            }
                            // get icon
                            PLVImageView plvImageView = (PLVImageView) findViewById(R.id.twImageView);
                            // get dp
                            int size = plvImageView.getWidth();
                            plvImageView.setUrl(iconUrl, size, size);
                        } catch (MalformedURLException e) {
                            Log.e("URLError", e.toString());
                        }

                        for (MediaEntity mediaEntity : mediaEntities) {
                            final String url = mediaEntity.getMediaURL();
                            addView(url);
                        }
                    }
                }

                public void addView(final String url) {
                    try {
                        LinearLayout baseLayout = (LinearLayout) findViewById(R.id.baseLayout);
                        // prev is last linear_layout
                        LinearLayout prevLayout = (LinearLayout) baseLayout.getChildAt(baseLayout.getChildCount() - 1);
                        LinearLayout currentLayout;
                        if (prevLayout.getChildCount() > 1) {
                            // make new linear_layout and put below prev
                            currentLayout = new LinearLayout(TwitterDisplay.this);
                            currentLayout.setLayoutParams(new AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
                            currentLayout.setOrientation(LinearLayout.HORIZONTAL);
                            baseLayout.addView(currentLayout);
                        } else {
                            // put new photo below prev photo (not new linear_layout)
                            currentLayout = prevLayout;
                        }
                        int width = baseLayout.getWidth();
                        // get dp
                        float dp = getResources().getDisplayMetrics().density;
                        // set padding and margin
                        int padding = (int) (2 * dp);
                        int margin = (int) (1 * dp);
                        // photo size
                        int size = width / 2 - padding * 2 - margin * 2;
                        // imgview size
                        int layoutsize = width / 2 - margin * 2;
                        // new imgview
                        PLVImageView imageView = new PLVImageView(TwitterDisplay.this);
                        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(layoutsize, layoutsize);
                        layoutParams.setMargins(margin, margin, margin, margin);
                        imageView.setLayoutParams(layoutParams);
                        imageView.setPadding(padding, padding, padding, padding);
                        imageView.setScaleType(ImageView.ScaleType.MATRIX);
                        // background (tap to gray)
                        imageView.setBackgroundResource(R.drawable.twitter_image_design);
                        imageView.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                // go to show fragment
                                Bundle bundle = new Bundle();
                                bundle.putString("url", url);
                                FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
                                ShowFragment showFragment = new ShowFragment();
                                showFragment.setArguments(bundle);
                                fragmentTransaction.replace(R.id.show_frag_replace, showFragment).commit();
                            }
                        });
                        imageView.setUrl(new URL(url), size, size);
                        currentLayout.addView(imageView);
                    } catch (MalformedURLException e) {
                        Log.e("URLError", e.toString());
                    }
                }
            });
        }

        @Override
        public void retweetedStatus(Status status) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(getApplicationContext(), getString(R.string.toast_retweet), Toast.LENGTH_LONG).show();
                }
            });

        }

        @Override
        public void createdFavorite(Status status) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(getApplicationContext(), getString(R.string.toast_favorite), Toast.LENGTH_LONG).show();
                }
            });

        }
    };

}

File: app/src/main/java/net/nonylene/photolinkviewer/TwitterDisplay.kt
package net.nonylene.photolinkviewer

import android.animation.LayoutTransition
import android.app.Activity
import android.app.Dialog
import android.app.DialogFragment
import android.app.Fragment
import android.content.Context
import android.content.Intent
import android.database.sqlite.SQLiteException
import android.os.Build
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AlertDialog
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import butterknife.bindView
import net.nonylene.photolinkviewer.fragment.*

import net.nonylene.photolinkviewer.tool.MyAsyncTwitter
import net.nonylene.photolinkviewer.tool.PLVUrl
import net.nonylene.photolinkviewer.tool.PLVUrlService
import net.nonylene.photolinkviewer.tool.ProgressBarListener
import net.nonylene.photolinkviewer.tool.TwitterStatusAdapter
import net.nonylene.photolinkviewer.view.HeightScalableScrollView
import net.nonylene.photolinkviewer.view.UserTweetLoadingView
import net.nonylene.photolinkviewer.view.UserTweetView

import java.util.regex.Pattern

import twitter4j.AsyncTwitter
import twitter4j.Status
import twitter4j.TwitterAdapter
import twitter4j.TwitterException
import twitter4j.TwitterMethod


class TwitterDisplay : Activity(), TwitterStatusAdapter.TwitterAdapterListener, ProgressBarListener, UserTweetView.TwitterViewListener, UserTweetLoadingView.LoadingViewListener {

    private var url: String? = null
    private var statusAdapter: TwitterStatusAdapter? = null
    private var twitter: AsyncTwitter? = null
    private var isSingle: Boolean = false
    private var isInitialized = false

    private val mTwitterSingleScrollView : HeightScalableScrollView by bindView(R.id.twitter_single_scroll)
    private val mTwitterSingleView: UserTweetView by bindView(R.id.twitter_single_view)
    private val mTwitterSingleLoadingView: UserTweetLoadingView by bindView(R.id.twitter_single_loading)
    private val mTwitterSingleDivider: View by bindView(R.id.twitter_single_divider)
    private val mProgressBar: ProgressBar by bindView(R.id.show_progress)
    private val mTweetBaseLayout: LinearLayout by bindView(R.id.tweet_base_view)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_twitter_display)

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1) {
            mTweetBaseLayout.layoutTransition.enableTransitionType(LayoutTransition.CHANGING)
        }

        if (Intent.ACTION_VIEW != intent.action) {
            Toast.makeText(this, "Intent Error!", Toast.LENGTH_LONG).show()
            return
        }

        val bundle = Bundle()
        url = intent.data.toString()
        bundle.putString("url", url)

        if (!getSharedPreferences("preference", Context.MODE_PRIVATE).getBoolean("authorized", false)) {
            Toast.makeText(applicationContext, getString(R.string.twitter_display_oauth), Toast.LENGTH_LONG).show()
            startActivity(Intent(this, TOAuth::class.java))
            finish()
            return
        }

        mTwitterSingleView.twitterViewListener = this
        mTwitterSingleView.visibility = View.GONE

        mTwitterSingleLoadingView.visibility = View.GONE
        mTwitterSingleLoadingView.loadingViewListener = this

        mTwitterSingleScrollView.isVerticalScrollBarEnabled = false

        val matcher = Pattern.compile("^https?://twitter\\.com/\\w+/status[es]*/(\\d+)").matcher(url)
        if (!matcher.find()) return
        val id_long = java.lang.Long.parseLong(matcher.group(1))

        try {
            twitter = MyAsyncTwitter.getAsyncTwitter(applicationContext).apply {
                addListener(twitterListener)
                showStatus(id_long)
            }
        } catch (e: SQLiteException) {
            e.printStackTrace()

        } catch (e: IndexOutOfBoundsException) {
            e.printStackTrace()
            Toast.makeText(applicationContext, getString(R.string.twitter_async_select), Toast.LENGTH_LONG).show()
            val intent = Intent(this, TOAuth::class.java)
            startActivity(intent)
            finish()
        }

        bundle.setTwitterEnabled(true)
        bundle.setTweetId(id_long)

        // option fragment
        val fragmentTransaction = fragmentManager.beginTransaction()
        val optionFragment = OptionFragment()
        optionFragment.arguments = bundle
        fragmentTransaction.add(R.id.root_layout, optionFragment).commit()
    }

    override fun onShowFragmentRequired(plvUrl: PLVUrl) {
        onFragmentRequired(ShowFragment(), plvUrl)
    }

    override fun onVideoShowFragmentRequired(plvUrl: PLVUrl) {
        onFragmentRequired(VideoShowFragment(), plvUrl)
    }

    private fun onFragmentRequired(fragment: Fragment, plvUrl: PLVUrl) {
        try {
            // go to show fragment
            val bundle = Bundle()
            bundle.putParcelable("plvurl", plvUrl)
            bundle.putBoolean("single_frag", isSingle)
            val fragmentTransaction = fragmentManager.beginTransaction()

            fragment.arguments = bundle
            fragmentTransaction.replace(R.id.show_frag_replace, fragment)

            // back to this screen when back pressed
            if (!isSingle) fragmentTransaction.addToBackStack(null)
            fragmentTransaction.commit()
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }

    }

    override fun onReadMoreClicked() {
        val replyId: Long

        if (statusAdapter != null) {
            replyId = statusAdapter!!.lastStatus!!.inReplyToStatusId
        } else {
            replyId = mTwitterSingleView.status!!.inReplyToStatusId
            mTwitterSingleScrollView.max = (resources.displayMetrics.heightPixels / 3.5f).toInt()
            mTwitterSingleScrollView.isVerticalScrollBarEnabled = true
            mTweetBaseLayout.setGravity(Gravity.CENTER_HORIZONTAL)
        }
        twitter!!.showStatus(replyId)
    }

    override fun hideProgressBar() {
        mProgressBar.visibility = View.GONE
    }

    class ChangeAccountDialog : DialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val accountsList = MyAsyncTwitter.getAccountsList(activity)
            val screen_list = accountsList.screenList
            val row_id_list = accountsList.rowIdList
            // get current screen_name
            val sharedPreferences = activity.getSharedPreferences("preference", Context.MODE_PRIVATE)
            val current_name = sharedPreferences.getString("screen_name", null)
            return AlertDialog.Builder(activity)
                    .setTitle(getString(R.string.account_dialog_title))
                    .setNegativeButton(getString(android.R.string.cancel), null)
                    .setSingleChoiceItems(screen_list.toTypedArray(), screen_list.indexOf(current_name),
                            { dialogInterface, position ->
                                // save rowid and screen name to preference
                                sharedPreferences.edit().putInt("account", row_id_list[position]).apply()
                                sharedPreferences.edit().putString("screen_name", screen_list[position]).apply()
                                //reload activity
                                startActivity(activity.intent)
                                dialog.dismiss()
                                activity.finish()
                            })
                    .create()
        }
    }

    private val twitterListener = object : TwitterAdapter() {

        override fun onException(e: TwitterException, twitterMethod: TwitterMethod?) {
            Log.e("twitterException", e.toString())
            val message = getString(R.string.twitter_error_toast) + ": " + e.statusCode + "\n(" + e.errorMessage + ")"
            when (e.errorCode) {
                179, 88 -> createDialog()
            }
            runOnUiThread { Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show() }
        }

        fun createDialog() {
            val accountDialog = ChangeAccountDialog()
            accountDialog.show(fragmentManager, "account")
        }

        override fun gotShowStatus(status: Status) {
            runOnUiThread{
                if (!isInitialized) {
                    isInitialized = true

                    // set media entity
                    val mediaEntities = status.extendedMediaEntities

                    // if number of media entity is one, show fragment directly
                    if (!PreferenceManager.getDefaultSharedPreferences(applicationContext).getBoolean("disp_tweet", false)
                            && (url!!.contains("/photo") || url!!.contains("/video")) && mediaEntities.size == 1) {
                        isSingle = true
                        val mediaEntity = mediaEntities[0]

                        if (mediaEntity.type in arrayOf("animated_gif", "video")) {
                            val plvUrl = PLVUrl(url)
                            // get biggest url
                            plvUrl.displayUrl = mediaEntity.videoVariants.filter {
                                ("video/mp4") == it.contentType
                            }.maxBy { it.bitrate }!!.url
                            onVideoShowFragmentRequired(plvUrl)

                        } else {
                            PLVUrlService(this@TwitterDisplay, object : PLVUrlService.PLVUrlListener {
                                override fun onGetPLVUrlFinished(plvUrls: Array<PLVUrl>) {
                                    val plvUrl = plvUrls[0]
                                    if (plvUrl.isVideo) onVideoShowFragmentRequired(plvUrl)
                                    else onShowFragmentRequired(plvUrl)
                                }

                                override fun onGetPLVUrlFailed(text: String) {}
                                override fun onURLAccepted() {}
                            }).requestGetPLVUrl(mediaEntity.mediaURLHttps)
                        }

                    } else {
                        hideProgressBar()
                        mTwitterSingleView.visibility = View.VISIBLE
                        mTwitterSingleView.setEntry(status)
                        if (status.inReplyToStatusId != (-1).toLong()) {
                            mTwitterSingleLoadingView.visibility = View.VISIBLE
                            mTwitterSingleDivider.visibility = View.VISIBLE
                        }

                    }
                } else {
                    if (statusAdapter == null) {
                        statusAdapter = TwitterStatusAdapter().apply {
                            twitterAdapterListener = this@TwitterDisplay
                        }
                        val listView = LayoutInflater.from(this@TwitterDisplay).inflate(R.layout.tweets_list_view, mTweetBaseLayout, false) as ListView
                        listView.adapter = statusAdapter
                        mTweetBaseLayout.addView(listView)
                        mTwitterSingleLoadingView.visibility = View.GONE
                        mTwitterSingleDivider.visibility = View.GONE
                    }
                    statusAdapter!!.addItem(status)
                }
            }
        }
    }
}
